src/lib.rs

```diff
@@ -213,10 +213,25 @@ pub fn run_compiled(
     let start_time = Instant::now();
 
     if let Some(mut hook) = hook {
-      while !root.redexes.is_empty() {
-        let readback = host.lock().unwrap().readback(root);
-        hook(&readback);
-        root.reduce(1);
+      if run_opts.lazy_mode {
```

Comment:
Can you separate each of these cases into separate functions?
So that our "run" function is simpler
```
match (hook, run_opts.lazy_mode, max_rwts) {
 (Some(hook), true, max_rwts) => normal_lazy_debug(),
 (...) => normal_strict_debug(),
 ...
}
```